core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotationUtil.java

```diff
@@ -404,26 +443,74 @@ private static void disallowedListAnnotation(Class<? extends Annotation> annotat
      */
     private static List<AnnotatedElement> resolveTargetElements(AnnotatedElement element,
                                                                 EnumSet<FindOption> findOptions) {
-        final List<AnnotatedElement> elements;
-        if (findOptions.contains(FindOption.LOOKUP_SUPER_CLASSES) && element instanceof Class) {
-            final Class<?> superclass = ((Class<?>) element).getSuperclass();
-            if ((superclass == null || superclass == Object.class) &&
-                ((Class<?>) element).getInterfaces().length == 0) {
-                elements = ImmutableList.of(element);
-            } else {
-                final Builder<AnnotatedElement> collector = new Builder<>();
-                collectSuperClasses((Class<?>) element, collector,
-                                    findOptions.contains(FindOption.COLLECT_SUPER_CLASSES_FIRST));
-                elements = collector.build();
-            }
+        if (!findOptions.contains(FindOption.LOOKUP_SUPER_CLASSES)) {
+            return ImmutableList.of(element);
+        }
+
+        final boolean collectSuperClassFirst = findOptions.contains(FindOption.COLLECT_SUPER_CLASSES_FIRST);
+
+        if (element instanceof Class) {
+            final List<Class<?>> classes = collectClasses((Class<?>) element, collectSuperClassFirst);
+            return convertToAnnotatedElements(classes);
+
+        } else if (element instanceof Method && findOptions.contains(FindOption.LOOKUP_SUPER_METHODS)) {
+            final Method method = (Method) element;
+            final Class<?> clazz = method.getDeclaringClass();
+            final List<Method> methods = collectMethods(clazz, method, collectSuperClassFirst);
+            return convertToAnnotatedElements(methods);
+
+        } else if (element instanceof Parameter && findOptions.contains(FindOption.LOOKUP_SUPER_PARAMETERS)) {
+            final Parameter parameter = (Parameter) element;
+            final Executable executable = parameter.getDeclaringExecutable();
+            final Class<?> clazz = executable.getDeclaringClass();
+            final List<Parameter> parameters = collectParameters(clazz, executable, parameter,
+                                                                 collectSuperClassFirst);
+            return convertToAnnotatedElements(parameters);
+
         } else {
-            elements = ImmutableList.of(element);
+            return ImmutableList.of(element);
+        }
+    }
+
+    private static <T extends AnnotatedElement> List<AnnotatedElement> convertToAnnotatedElements(
+            List<T> elements) {
+        return elements.stream().map(element -> (AnnotatedElement)element).collect(Collectors.toList());
```

Comment:
We prefer using immutable list.
```suggestion
        return elements.stream().map(element -> (AnnotatedElement) element).collect(toImmutableList());
```